File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/DiscreteKalmanFilter.cs
using System;
public class DiscreteKalmanFilter<TState, TMeasurement> : BaseKalmanFilter<TState, TMeasurement>
{
    // Methods
    public DiscreteKalmanFilter(TState initialState, float[,] initialStateError, int measurementVectorDimension, int controlVectorDimension, Func<TState, float[]> stateConvertFunc, Func<float[], TState> stateConvertBackFunc, Func<TMeasurement, float[]> measurementConvertFunc) : base(initialState, initialStateError, measurementVectorDimension, controlVectorDimension, stateConvertFunc, stateConvertBackFunc, measurementConvertFunc)
    {
    }

    private void correct(float[] innovationVector)
    {
        if (innovationVector.Length != base.MeasurementVectorDimension)
        {
            throw new Exception("PredicitionError error vector (innovation vector) must have the same length as measurement.");
        }
        base.state = base.state.Add(base.KalmanGain.Multiply(innovationVector));
        float[,] a = MatrixF.Identity(base.StateVectorDimension);
        base.EstimateCovariance = a.Subtract(base.KalmanGain.Multiply(base.MeasurementMatrix), false).Multiply(base.EstimateCovariance.Transpose<float>());
    }

    protected override void CorrectInternal(float[] measurement)
    {
        float[] innovationVector = base.CalculateDelta(measurement);
        this.correct(innovationVector);
    }

    protected override void predictInternal(float[] controlVector)
    {
        base.state = base.TransitionMatrix.Multiply(base.state);
        if (controlVector != null)
        {
            base.state = base.state.Add(base.ControlMatrix.Multiply(controlVector));
        }
        base.EstimateCovariance = base.TransitionMatrix.Multiply(base.EstimateCovariance).Multiply(base.TransitionMatrix.Transpose<float>()).Add(base.ProcessNoise);
        float[,] b = base.MeasurementMatrix.Transpose<float>();
        base.ResidualCovariance = base.MeasurementMatrix.Multiply(base.EstimateCovariance).Multiply(b).Add(base.MeasurementNoise);
        base.ResidualCovarianceInv = base.ResidualCovariance.Inverse();
        base.KalmanGain = base.EstimateCovariance.Multiply(b).Multiply(base.ResidualCovarianceInv);
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/DiscreteKalmanFilter.swift
//
//  DiscreteKalmanFilter.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/8.
//

import Foundation

class DiscreteKalmanFilter<TState, TMeasurement>: BaseKalmanFilter<TState, TMeasurement> {
    required init(
        initialState: TState,
        initialStateError: [[Float]],
        measurementVectorDimension: Int,
        controlVectorDimension: Int,
        stateConvertFunc: @escaping StateConvertFunc,
        stateConvertBackFunc: @escaping StateConvertBackFunc,
        measurementConvertFunc: @escaping MeasurementConvertFunc
    ) {
        super.init(
            initialState: initialState,
            initialStateError: initialStateError,
            measurementVectorDimension: measurementVectorDimension,
            controlVectorDimension: controlVectorDimension,
            stateConvertFunc: stateConvertFunc,
            stateConvertBackFunc: stateConvertBackFunc,
            measurementConvertFunc: measurementConvertFunc
        )
    }
    
    func correct(innovationVector: [Float]) {
        if innovationVector.count != measurementVectorDimension {
            fatalError("PredicitionError error vector (innovation vector) must have the same length as measurement.")
        }
        
        // TODO: 验证、优化强制解包逻辑
        let b = VectorF.mul(kalmanGain!, innovationVector)
        state = VectorF.add(state!, b)
    }
    
    // MARK: - From Base
    override func correctInternal(measurement: [Float]) {
        let innovationVector = calculateDelta(measurement: measurement)
        correct(innovationVector: innovationVector)
    }
    
    override func predictInternal(controlVector: [Float]?) {
        // TODO: 验证、优化强制解包逻辑
        
        state = VectorF.mul(transitionMatrix!, state!)

        if let controlVector = controlVector {
            state = VectorF.add(
                state!,
                VectorF.mul(controlMatrix!, controlVector)
            )
        }
        
        estimateCovariance = MatrixF.add(
            MatrixF.mul(
                MatrixF.mul(transitionMatrix!, estimateCovariance!),
                MatrixF.transpose(transitionMatrix!)
            ),
            processNoise!
        )
        
        let b = MatrixF.transpose(measurementMatrix!)
        residualCovariance = MatrixF.add(
            MatrixF.mul(
                MatrixF.mul(measurementMatrix!, estimateCovariance!),
                b
            ),
            measurementNoise!
        )

        residualCovarianceInv = MatrixF.inverse(residualCovariance!)
        kalmanGain = MatrixF.mul(
            MatrixF.mul(estimateCovariance!, b),
            residualCovarianceInv!
        )
    }
}
